[KEP-3521] Part 2: Core scheduling implementation

Huang-Wei opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature
/sig scheduling

#### What this PR does / why we need it:

This PR is rebased atop Part 1 (#113274), covering the following logic:

- [x] API, Codegen, UT for Enqueue extension point
- [x] Add EnqueuePlugin to Scheduler Framework
- [x] Implement EnqueuePluign in scheduler queue
- [x] Implementation of DefaultEnqueue plugin
- [x] Metrics

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of #113269

#### Special notes for your reviewer:


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
A new `preEnqueue` extension point is added to scheduler's component config v1beta2/v1beta3/v1. 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3521-pod-scheduling-readiness
```

Huang-Wei commented:
cc @ahg-g this is the 2nd PR of the overall KEP impl. There is a third PR upcoming to cover integration/e2e tests.

Huang-Wei commented:
/retest

Huang-Wei commented:
/retest

ahg-g commented on pkg/scheduler/metrics/metrics.go:
```diff
@@ -92,7 +92,7 @@ var (
 			Subsystem:      SchedulerSubsystem,
 			Name:           "pending_pods",
-			Help:           "Number of pending pods, by the queue type. 'active' means number of pods in activeQ; 'backoff' means number of pods in backoffQ; 'unschedulable' means number of pods in unschedulablePods.",
+			Help:           "Number of pending pods, by the queue type. 'active' means number of pods in activeQ; 'backoff' means number of pods in backoffQ; 'unschedulable' means number of pods in unschedulablePods that have been tried scheduling at least once; 'suspended' means number of pods in unschedulablePods that are suspended by enqueue plugins.",
```
"that the scheduler attempted to schedule and failed; gated is the number of unschedulable pods that the scheduler never attempted to schedule because they are gated"

Huang-Wei replied:
done.

ahg-g commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -318,12 +320,12 @@ func (p *PriorityQueue) Run() {
-
-	p.unschedulablePods.addOrUpdate(pInfo)
-	return false
+	pInfo.Suspended = !success
```
`gated`

ahg-g commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -332,13 +334,23 @@ func (p *PriorityQueue) PreEnqueue(pInfo *framework.QueuedPodInfo) bool {
 			continue
 		}
+		if pInfo.UnschedulablePlugins == nil {
+			pInfo.UnschedulablePlugins = sets.NewString()
```
shouldn't this be initialized on pInfo creation?

Huang-Wei replied:
yes, it was for adapting to a ut. Updated the UT now.

leilajal commented:
/remove-sig api-machinery

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -300,12 +314,47 @@ func (p *PriorityQueue) Run() {
 	pInfo := p.newQueuedPodInfo(pod)
+	if !p.PreEnqueue(pInfo) {
+		return nil
+	}
```
sg, it's just we need to distinguish between successful p.PreEnqueue and p.activeQ.Add. Updated.

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -300,12 +314,47 @@ func (p *PriorityQueue) Run() {
+			continue
+		}
+		pInfo.UnschedulablePlugins.Insert(pl.Name())
+		klog.InfoS("PreEnqueue failed", "pod", klog.KObj(pod), "plugin", pl.Name(), "error", s.AsError())
```
I'd do `klog.V(5).InfoS` on failures and `klog.ErrorS` on unexpected errors, which is consistent with what we did in other extension point:

https://github.com/kubernetes/kubernetes/blob/a0b69ecd01edc68f9eb88658edcb9f82daf27883/pkg/scheduler/schedule_one.go#L145-L150

Huang-Wei commented on pkg/scheduler/framework/plugins/defaultenqueue/default_enqueue.go:
```diff
@@ -0,0 +1,67 @@
+limitations under the License.
+*/
+
+package defaultenqueue
```
sg, renamed the folder (defaultenqueue -> enqueue) as well.

ahg-g commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -346,28 +347,33 @@ func (p *PriorityQueue) runEnqueuePlugins(ctx context.Context, pInfo *framework.
 
+func (p *PriorityQueue) addToActiveQ(pInfo *framework.QueuedPodInfo) error {
+	if !p.PreEnqueue(pInfo) {
+		return errors.New("preEnqueue checks failed")
```
but it is not an error if PreEnqueue returns false, it just means the pod is gated

ahg-g commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -346,28 +347,33 @@ func (p *PriorityQueue) runEnqueuePlugins(ctx context.Context, pInfo *framework.
+	if !p.PreEnqueue(pInfo) {
+		return errors.New("preEnqueue checks failed")
+	}
+	return p.activeQ.Add(pInfo)
```
log error here on failure, and return boolean to indicate if the pod was added

Huang-Wei replied:
It's neat. Thanks!

ahg-g commented on pkg/scheduler/framework/plugins/enqueue/scheduling_gates.go:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 */
 
-package defaultenqueue
+package enqueue
```
I would name the folder and pkg `schedulinggates` too, just like other features. Remember we could have other enqueue plugins, just like `preFilter` etc.

Huang-Wei commented on pkg/scheduler/scheduler.go:
```diff
@@ -317,6 +321,7 @@ func New(client clientset.Interface,
 		internalqueue.WithPodNominator(nominator),
 		internalqueue.WithClusterEventMap(clusterEventMap),
 		internalqueue.WithPodMaxInUnschedulablePodsDuration(options.podMaxInUnschedulablePodsDuration),
+		internalqueue.WithEnqueuePluginMap(enqueuePluginMap),
```
when running preEnqueue plugins, without a map, how can you tell which plugin set is associated with that Pod (via its `schedulerName`)?

QueueSortFunc is using a global sort function. Enqueue plugins are profile specific.

Huang-Wei commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -446,6 +446,69 @@ func TestPriorityQueue_Activate(t *testing.T) {
+}
+
+func TestPriorityQueue_PreEnqueue(t *testing.T) {
+	tests := []struct {
```
> I think we should cover testcases with one more enqueuePlugins

SG.

> also test the unschedulable plugins.

Nope, that's covered in other UTs or integration test.

Huang-Wei commented:
@kerthcet @ahg-g thanks for the review. Comments addressed, PTAL.

ahg-g commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -448,15 +448,19 @@ func TestPriorityQueue_Activate(t *testing.T) {
 
-type preEnqueuePlugin struct{}
+type preEnqueuePlugin struct {
+	whitelists []string
```
allowlists

ahg-g commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -300,19 +317,75 @@ func (p *PriorityQueue) Run() {
+// PreEnqueue firstly runs registered EnqueuePlugins, and returns true if all plugins run
+// successfully; otherwise add or update the pod in `unschedulablePods`, and returns false.
+func (p *PriorityQueue) PreEnqueue(pInfo *framework.QueuedPodInfo) bool {
+	success := p.runEnqueuePlugins(context.Background(), pInfo)
+	if !success {
+		// Add the Pod to unschedulablePods if it's not passing EnqueuePlugins.
+		p.unschedulablePods.addOrUpdate(pInfo)
+	}
+	pInfo.Gated = !success
+	return success
```
I find this function redundant, can we just move this logic into `addToActiveQ`, see the comment on that function.

ahg-g replied:
also, why are we exporting it?

ahg-g commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -300,19 +317,75 @@ func (p *PriorityQueue) Run() {
+// upon the first failure.
+// Note: we need to associate the failed plugin to `pInfo`, so that the pod can be moved back
+// to activeQ by related cluster event.
+func (p *PriorityQueue) runEnqueuePlugins(ctx context.Context, pInfo *framework.QueuedPodInfo) bool {
```
should we move this function to the framework just like all run*Plugins functions?

Huang-Wei replied:
nope, `frameworkImpl` doesn't hold scheudlingqueue, and hence cannot perform `runEnqueuePlugins`.

ahg-g commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -300,19 +317,75 @@ func (p *PriorityQueue) Run() {
+// 1. a boolean flag to indicate whether the pod is added successfully.
+// 2. an error for the caller to act on.
+func (p *PriorityQueue) addToActiveQ(pInfo *framework.QueuedPodInfo) (bool, error) {
+	if !p.PreEnqueue(pInfo) {
+		return false, nil
+	}
```
```suggestion
       pInfo.Gated = !p.runEnqueuePlugins(context.Background(), pInfo)
       if p.Info.Gated {
           // Add the Pod to unschedulablePods if it's not passing EnqueuePlugins.
           p.unschedulablePods.addOrUpdate(pInfo)
	   return false, nil
	}
```

Huang-Wei replied:
sg, updated.

ahg-g commented:
/label api-review

For the changes to component config in staging

ahg-g commented on pkg/api/pod/util_test.go:
```diff
@@ -1935,3 +1935,85 @@ func TestDropHostUsers(t *testing.T) {
+					}
+					dropDisabledFields(&newPod.Spec, nil, oldPodSpec, nil)
+					// Old Pod should never be changed.
+					if !reflect.DeepEqual(oldPod, oldPodInfo.pod()) {
```
sorry, just noticed this, why not just use `cmp.Diff`? you're already calling it for the error message below.

Huang-Wei replied:
This needs to be polished indeed. (the snippet was copied over from somewhere) I will update in the precedent PR and rebase here.

ahg-g commented:
@liggitt this one needs a quick API review to few additions to the component config scheduler API. The pod API review is done under https://github.com/kubernetes/kubernetes/pull/113274

Huang-Wei commented:
/retest

denkensk commented on pkg/scheduler/internal/queue/scheduling_queue_test.go:
```diff
@@ -446,6 +446,73 @@ func TestPriorityQueue_Activate(t *testing.T) {
+}
+
+func (pl *preEnqueuePlugin) PreEnqueue(ctx context.Context, p *v1.Pod) *framework.Status {
+	for _, allowed := range append(pl.allowlists, queueable) {
```
I am not sure if i understand correctly：maybe we can delete the default allow name `queueable` to make the test understood easily.
And change the case from
```go
  plugins:  []framework.EnqueuePlugin{&preEnqueuePlugin{allowlists: []string{"foo"}}, &preEnqueuePlugin{}},
- pod:      st.MakePod().Name(queueable).Obj(),
+ pod:      st.MakePod().Name("foo").Obj(),
```

Huang-Wei replied:
udpated.

denkensk commented on pkg/scheduler/metrics/metrics.go:
```diff
@@ -92,7 +92,7 @@ var (
 			Subsystem:      SchedulerSubsystem,
 			Name:           "pending_pods",
-			Help:           "Number of pending pods, by the queue type. 'active' means number of pods in activeQ; 'backoff' means number of pods in backoffQ; 'unschedulable' means number of pods in unschedulablePods.",
+			Help:           "Number of pending pods, by the queue type. 'active' means number of pods in activeQ; 'backoff' means number of pods in backoffQ; 'unschedulable' means number of pods in unschedulablePods that the scheduler attempted to schedule and failed; 'gated' is the number of unschedulable pods that the scheduler never attempted to schedule because they are gated.",
```
I find you will run `runEnqueuePlugins` in `flushBackoffQCompleted()`. These pods in `flushBackoffQCompleted()` which are gated are not the pods never attempted to schedule, right?

denkensk replied:
Got it by you demo show, the gated just can be added when creating the pod.

kerthcet commented on pkg/scheduler/framework/plugins/schedulinggates/scheduling_gates.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+var _ framework.EnqueuePlugin = &SchedulingGates{}
+var _ framework.EnqueueExtensions = &SchedulingGates{}
```
Not related to this PR, but the name is a little confusing.

Huang-Wei replied:
Do you mean the plugin name as `SchedulingGates`?

kerthcet replied:
The name of `EnqueueExtensions`, we have `ScoreExtensions` for `ScorePlugin` for example, but the `EnqueueExtensions` is not related to plugins.

ahg-g replied:
also, should we name the plugin itself `PreEnqueue` just like PreFilter/PreScore/PreBind?

Huang-Wei replied:
> The name of EnqueueExtensions

Yup, it's a bit misleading as `EnqueueExtensions` is supposed to be implemented by any plugin that implements a filter-like extension point - PreFitler, Filter, Enqueue.

 > should we name the plugin itself PreEnqueue

you mean the extension point right? we may do that once we expose queuing logic as an extension point, and then rename the current one as PreEnqueue. For now, the fact is we only expose PreEnqueue in this extension point, and the core queuing logic along with the data structure is not exposed.

ahg-g replied:
> you mean the extension point right? we may do that once we expose queuing logic as an extension point, 

Harder to do that later because it will break configurations that explicitly set/unset schedulingGates.

Huang-Wei replied:
then let's rename it to PreEnqueue. Updated.

kerthcet commented on pkg/scheduler/framework/plugins/schedulinggates/scheduling_gates.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+var _ framework.EnqueuePlugin = &SchedulingGates{}
+var _ framework.EnqueueExtensions = &SchedulingGates{}
```
Not related to this PR, but the name is a little confusing now.

logicalhan commented:
/assign
/triage accepted

kerthcet commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -560,13 +622,13 @@ func (p *PriorityQueue) Update(oldPod, newPod *v1.Pod) error {
 				klog.V(5).InfoS("Pod moved to an internal scheduling queue", "pod", klog.KObj(pInfo.Pod), "event", PodUpdate, "queue", backoffQName)
 			} else {
-				if err := p.activeQ.Add(pInfo); err != nil {
+				if added, err := p.addToActiveQ(pInfo); !added {
```
This should be `err != nil`

Huang-Wei replied:
nope, `!added` indicates either pod is gated, or error in adding pod to activeQ, so we should return in these cases.

`err != nil` missed the case that pod is gated.

kerthcet replied:
Make sense. Then if pod is gated, we should always return false in `isPodBackingoff`, right?  Or when backoff time arrives, we'll move pod to backoffQ.

kerthcet replied:
Another approach is similar to activeQ, we'll wrap the function of `p.podBackoffQ.Add()`, when it's pod gated, we'll move to unschedulablePods.

kerthcet replied:
What about this? @Huang-Wei

Huang-Wei replied:
Per offline discussion with @kerthcet, this is a very good point. And with current logic, it's problematic that a pod may be put to backoffQ in the following scenarios:

- pod's scheduling gates are removed within `podInitialBackoffSeconds` - in this case, isPodBackoff returns true and hence when a pod is updated, the pod is put into backoffQ
- when `movePodsToActiveOrBackoffQueue` is called (e.g., flushUnschedulablePodsLeftover or Node added), isPodBackingoff is also called and may returns true and move a gated pod to backoffQ.

I will append a fix in the next PR.

kerthcet commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -579,7 +641,7 @@ func (p *PriorityQueue) Update(oldPod, newPod *v1.Pod) error {
 	// If pod is not in any of the queues, we put it in the active queue.
 	pInfo := p.newQueuedPodInfo(newPod)
-	if err := p.activeQ.Add(pInfo); err != nil {
+	if added, err := p.addToActiveQ(pInfo); !added {
```
ditto.

kerthcet commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -806,25 +867,33 @@ type UnschedulablePods struct {
-	// metricRecorder updates the counter when elements of an unschedulablePodsMap
-	// get added or removed, and it does nothing if it's nil
-	metricRecorder metrics.MetricRecorder
+	// metricRecorder/gatedRecorder updates the counter when elements of an unschedulablePodsMap
```
No metricRecorder now

kerthcet replied:
And this seems like it will impact current metrics now.

Huang-Wei replied:
> No metricRecorder now

updated.

> And it seems like will impact current metrics now.

Metric's sub-label addition should be ok I suppose, as only users adopting the new feature will use the metric with new label; otherwise, all unschedulable pods will still fall into the metric with old labels.

kerthcet reviewed (commented):
Some nits.

logicalhan reviewed (approved):
/lgtm
/approve
(for the stable metrics HELP changes, which is fine).

kerthcet commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -446,19 +506,21 @@ func (p *PriorityQueue) flushBackoffQCompleted() {
-		klog.V(5).InfoS("Pod moved to an internal scheduling queue", "pod", klog.KObj(pod), "event", BackoffComplete, "queue", activeQName)
-		metrics.SchedulerQueueIncomingPods.WithLabelValues("active", BackoffComplete).Inc()
-		activated = true
+		if added, _ := p.addToActiveQ(pInfo); added {
```
Will gated pod possible in backoffQ? Related to the question I asked here https://github.com/kubernetes/kubernetes/pull/113275/files#r1013651169

Huang-Wei replied:
It's technically impossible for a gated pod to stay in backoffQ.

kerthcet replied:
Then `p.activeQ.Add` seems enough. For we only support one-direction state transition from gated -> non-gated, and pods in backoffQ are all non-gated.

Huang-Wei replied:
that's a fair point. Let me record it as a follow-up in #113608.

ahg-g commented:
/milestone v1.26

liggitt commented on staging/src/k8s.io/kube-scheduler/config/v1/types.go:
```diff
@@ -170,6 +170,9 @@ type KubeSchedulerProfile struct {
 // be invoked before default plugins, default plugins must be disabled and re-enabled here in desired order.
 type Plugins struct {
+	// Enqueue is a list of plugins that should be invoked before adding pods to the scheduling queue.
+	Enqueue PluginSet `json:"enqueue,omitempty"`
```
is this a new capability? if so, does putting a value in this config field require a feature gate to be turned on?

liggitt replied:
if so, docs need to indicate the feature gate connected to this

ahg-g replied:
We never used feature flags when adding new extensions since for the most part we consider them a refactor and a no-op. The schedulingGates feature (which is implemented using this new extension of the scheduler framework) itself is behind a feature gate.

Huang-Wei replied:
no, we don't disable the extension point itself no matter the featuregate is disabled or not. It's documented intentionally as a goal in the KEP for scheduler developers to use from day 1.

liggitt commented on pkg/scheduler/apis/config/v1/default_plugins.go:
```diff
@@ -29,6 +29,7 @@ func getDefaultPlugins() *v1.Plugins {
 	plugins := &v1.Plugins{
 		MultiPoint: v1.PluginSet{
 			Enabled: []v1.Plugin{
+				{Name: names.SchedulingGates},
```
is SchedulingGates a new plugin? is it behind a feature gate? if so, should this be considering whether the feature is enabled or not when including this or is the gating accomplished within the plugin itself?

put another way, if a user explicitly included the new alpha SchedulingGates plugin in their config without enabling the feature gate, would the config load and run successfully with that plugin inert, and then break if we ended up needing to drop the alpha feature and deleted the SchedulingGates plugin?

ahg-g replied:
+1, adding the plugin here should be predicated on the feature gate

Huang-Wei replied:
It's a new plugin and always enabled, but it _functions_ behind a feature gate.

```go
func (pl *SchedulingGates) PreEnqueue(ctx context.Context, p *v1.Pod) *framework.Status {
	if !pl.enablePodSchedulingReadiness || len(p.Spec.SchedulingGates) == 0 {
		return nil
	}
...
```

> +1, adding the plugin here should be predicated on the feature gate

Actually no, like VolumeBinding's Score plugin is always enabled, but it functions behind the feature gate:

https://github.com/kubernetes/kubernetes/blob/abcbaed0784baf5ed2382aae9705a8918f2daa18/pkg/scheduler/framework/plugins/volumebinding/volume_binding.go#L380-L391

ahg-g replied:
> Actually no, like VolumeBinding's Score plugin is always enabled, but it functions behind the feature gate:

@Huang-Wei I think this was bad precedence, it is safer to gate the addition of the plugin to avoid breaking configurations that explicitly add the plugin in case the feature doesn't make it beyond alpha for some reason.

Huang-Wei replied:
ok, I just updated the logic to enable ScheudlingGate when the feature gate is enabled. But we're inconsistent in VolumeBinding plugin, will submit a separate PR fixing this.

ahg-g replied:
Looks good to me, @liggitt ?

Huang-Wei replied:
Created an issue to track the VolumeBinding issue: https://github.com/kubernetes/kubernetes/issues/113705

liggitt commented on pkg/scheduler/apis/config/v1beta2/zz_generated.conversion.go:
```diff
@@ -761,6 +761,7 @@ func Convert_v1beta2_Plugins_To_config_Plugins(in *v1beta2.Plugins, out *config.
 }
 
 func autoConvert_config_Plugins_To_v1beta2_Plugins(in *config.Plugins, out *v1beta2.Plugins, s conversion.Scope) error {
+	// WARNING: in.Enqueue requires manual conversion: does not exist in peer-type
```
flagging this to make sure it was intentional this field was only added to the v1 type. The release note should be clear about what field is added and to which versions, and/or link to docs for the feature config

ahg-g replied:
Adding it to previous existing versions should be ok. @Huang-Wei wdyt?

liggitt replied:
you don't have to, if you want to use this as a carrot to incentivize upgrading to v1 config, that can be fine. I just wanted to make sure it was an intentional decision

liggitt replied:
though I would ask, if someone uses a v1beta1 default config, is the behavior different than a v1 default config?

ahg-g replied:
yes if the schedulingGates feature flag is enabled, so we have to add it.

Huang-Wei replied:
ok, let me add it to all supported versions.

liggitt commented on pkg/scheduler/apis/config/v1beta3/zz_generated.conversion.go:
```diff
@@ -751,6 +751,7 @@ func Convert_v1beta3_Plugins_To_config_Plugins(in *v1beta3.Plugins, out *config.
 }
 
 func autoConvert_config_Plugins_To_v1beta3_Plugins(in *config.Plugins, out *v1beta3.Plugins, s conversion.Scope) error {
+	// WARNING: in.Enqueue requires manual conversion: does not exist in peer-type
```
flagging this to make sure it was intentional this field was only added to the v1 type. The release note should be clear about what field is added and to which versions, and/or link to docs for the feature config

liggitt commented on pkg/scheduler/apis/config/types.go:
```diff
@@ -138,6 +138,9 @@ type KubeSchedulerProfile struct {
 // be invoked before default plugins, default plugins must be disabled and re-enabled here in desired order.
 type Plugins struct {
+	// Enqueue is a list of plugins that should be invoked before adding pods to the scheduling queue.
+	Enqueue PluginSet
```
I expected to see changes to pkg/scheduler/apis/config/validation/validation.go for this field, both that it is a valid pluginSet, and also that the field is empty when the new feature gate is turned off

ahg-g replied:
We need to add the new extension to https://github.com/kubernetes/kubernetes/blob/abcbaed0784baf5ed2382aae9705a8918f2daa18/pkg/scheduler/apis/config/validation/validation.go#L200, but it will be a no-op as of now. 

> and also that the field is empty when the new feature gate is turned off

The feature gate is for the schedulingGates feature, not the extension. I am not sure if we want to gate the extension itself.

liggitt reviewed (commented):
field name and types.go change looks good, a few questions about validation / feature gating and making sure that it was intentional to only add to the v1 config

ahg-g commented on pkg/scheduler/apis/config/types.go:
```diff
@@ -138,6 +138,9 @@ type KubeSchedulerProfile struct {
 // Enabled plugins are called in the order specified here, after default plugins. If they need to
 // be invoked before default plugins, default plugins must be disabled and re-enabled here in desired order.
 type Plugins struct {
+	// Enqueue is a list of plugins that should be invoked before adding pods to the scheduling queue.
```
we need to add the new extension to https://github.com/kubernetes/kubernetes/blob/abcbaed0784baf5ed2382aae9705a8918f2daa18/pkg/scheduler/apis/config/types.go#L229

Huang-Wei replied:
done.

sanposhiho commented on pkg/scheduler/apis/config/types.go:
```diff
@@ -138,6 +138,9 @@ type KubeSchedulerProfile struct {
 // Enabled plugins are called in the order specified here, after default plugins. If they need to
 // be invoked before default plugins, default plugins must be disabled and re-enabled here in desired order.
 type Plugins struct {
+	// PreEnqueue is a list of plugins that should be invoked before adding pods to the scheduling queue.
```
Need to specify which queue of the scheduling queue to enter.

```suggestion
	// PreEnqueue is a list of plugins that should be invoked before adding pods to the activeQ in the scheduling queue.
```

Huang-Wei replied:
probably not worthing indicating the details in the comments. (the comments of QueueSort also says "in the scheduling queue")

sanposhiho commented on pkg/scheduler/internal/queue/scheduling_queue.go:
```diff
@@ -234,6 +240,13 @@ func WithPodMaxInUnschedulablePodsDuration(duration time.Duration) Option {
 	}
 }
 
+// WithEnqueuePluginMap sets preEnqueuePluginMap for PriorityQueue.
+func WithEnqueuePluginMap(m map[string][]framework.PreEnqueuePlugin) Option {
```
You might want to rename it to `WithPreEnqueuePluginMap` because we named the interface `PreEnqueue`.

Huang-Wei replied:
done.

sanposhiho commented on pkg/scheduler/scheduler.go:
```diff
@@ -309,6 +309,10 @@ func New(client clientset.Interface,
 		return nil, errors.New("at least one profile is required")
 	}
 
+	enqueuePluginMap := make(map[string][]framework.PreEnqueuePlugin)
```
Should we move the PreEnqueue extension point itself behind the feature gate? How can users disable the PreEnqueue extension point if the PreEnqueue extension point itself (not `.spec.schedulingGate` or the scheduling gate plugin) has the bug.
(Please ignore if this topic is already discussed. I haven't followed all discussions 🙏 )

Huang-Wei replied:
See https://github.com/kubernetes/kubernetes/pull/113275#discussion_r1014341424.

liggitt commented on pkg/scheduler/apis/config/validation/validation.go:
```diff
@@ -198,6 +198,7 @@ func validatePluginConfig(path *field.Path, apiVersion string, profile *config.K
 
 	if profile.Plugins != nil {
 		stagesToPluginSet := map[string]config.PluginSet{
+			"enqueue":    profile.Plugins.PreEnqueue,
```
update to match the new field name
```suggestion
			"preEnqueue":    profile.Plugins.PreEnqueue,
```

Huang-Wei replied:
Updated, and with more literal replacements.

liggitt reviewed (commented):
one comment on the validation, API bits lgtm otherwise

liggitt commented:
/approve
for API type changes

/hold for validation fieldname change in https://github.com/kubernetes/kubernetes/pull/113275#discussion_r1015889541

scheduling reviewers can lgtm and unhold once that is done

Huang-Wei commented:
@ahg-g almost all comments have been addressed. Let me know if it's good to squash.

ahg-g commented:
> @ahg-g almost all comments have been addressed. Let me know if it's good to squash.

looks good, please squash

Huang-Wei committed:
API, Codegen, UT for PreEnqueue extension point

Huang-Wei committed:
PreEnqueue implementation

- Add PreEnqueuePlugin to Scheduler Framework
- Implement PreEnqueuePlugin in scheduler queue
- Implementation of SchedulingGates plugin
- Metrics

Huang-Wei commented:
/retest

Huang-Wei commented:
@ahg-g commits squashed and CI is green now.

ahg-g commented:
/hold cancel
/lgtm

BinacsLee commented:
/retest

pacoxu commented:
/retest
